### PR TITLE
[feature] Project Log Feed [OSF-5980] [OSF-6454] [OSF-4582] [OSF-6386]

### PR DESCRIPTION
**Pull request by caseyrollins**
#### Purpose
- Updates the Recent Activity widget on the project view page to use the new log feed that appears on the My Projects page. Additionally, the logs for the project and its components are now retrieved from API v2 endpoints. 

#### Changes
- Updated log feed UI. 
- Implemented pagination on the project page log feed.
- Logs on the view project page retrieved from API v2. 
- Added a specific log text file for anonymized logs.

#### Side Effects
- The 'My Projects' page and its Activity Logs widget should be unaffected by this PR, but the page should still be tested as a minor change was made to the file that renders the log text.
- This PR makes the OSF rich in dead log code (which could be removed now in this PR, or later as part of [a maintenance/improvement JIRA ticket](https://openscience.atlassian.net/browse/OSF-6491)). 

#### QA Notes
- Everywhere on the OSF where logs for a project/component/registration are shown should be tested. Noticeable errors that would definitely be related to this PR would include receiving an endless loading icon or an 'Unable to retrieve logs' message -- there would also be console errors in both of these cases. 
- Places to check:
 - project overview page (for both projects and registrations -- with and without anonymous VOLs)
 - components widget on the project overview page, the 'Recent Activity' dropdown for components (public and private) and links (public and private)
 - profile page, 'Recent Activity' dropdown for listed public components and public projects
 - registrations tab for a project, the 'Recent Activity' dropdown for the listed registrations (pending, archiving, withdrawn)

#### Tickets
- [OSF-5980](https://openscience.atlassian.net/browse/OSF-5980)
- [OSF-6454](https://openscience.atlassian.net/browse/OSF-6454)
- [OSF-4582](https://openscience.atlassian.net/browse/OSF-4582)
- [OSF-6386](https://openscience.atlassian.net/browse/OSF-6386)

#### Screenshots
**- Recent Activity -- Project -- Before**

screen shot 2016-06-10 at 2 11 24 pm

**- Recent Activity -- Project -- After**

screen shot 2016-06-15 at 10 52 51 am

**- Recent Activity -- Component -- Before**
screen shot 2016-06-10 at 2 12 12 pm

**- Recent Activity -- Component -- After**
screen shot 2016-06-15 at 10 53 06 am

**- Recent Activity -- Anonymous VOL -- Before**

screen shot 2016-06-10 at 2 12 28 pm

**- Recent Activity -- Anonymous VOL -- After**

screen shot 2016-06-15 at 10 53 18 am

**- Institution Log -- Anonymous VOL -- Before**

screen shot 2016-06-10 at 2 12 35 pm

**- Institution Log -- Anonymous VOL -- After**

screen shot 2016-06-10 at 2 07 57 pm

### CODE REVIEW

File: website/static/js/logFeed.js [hunk 1]
```diff
@@ -1,238 +1,94 @@
...
+        return m('.db-activity-list.m-t-md', [
+            ctrl.activityLogs() ? ctrl.activityLogs().map(function(item){
+                if (ctrl.node.anonymous) { item.anonymous = true; }
+                var image = m('i.fa.fa-desktop');
+                if (!item.anonymous && item.embeds.user && item.embeds.user.data) {
+                    image = m('img', { src : item.embeds.user.data.links.profile_image});
+                } else if (!item.anonymous && item.embeds.user && item.embeds.user.errors[0].meta){
+                    image = m('img', { src : item.embeds.user.errors[0].meta.profile_image});
+                }
+                return m('.db-activity-item', [
```
Comment (mfraezz) [inline website/static/js/logFeed.js]: This indentation here is difficult to read, please follow the [cosdev JS style](http://cosdev.readthedocs.io/en/latest/style_guides/javascript.html)

File: website/static/js/_anonymousLogTexts.json [hunk 2]
```diff
@@ -0,0 +1 @@
...user changed permissions for a project","s3_bucket_linked":"A user linked an Amazon S3 bucket to a project","s3_bucket_unlinked":"A user unselected an Amazon S3 bucket in a project","s3_file_added":"A user added a file to an Amazon S3 bucket in a project","s3_file_removed":"A user removed a file in an Amazon S3 bucket in a project","s3_file_updated":"A user updated a file in an Amazon S3 bucket in a project","s3_folder_created":"A user created a folder in an Amazon S3 in a project","s3_node_authorized":"A user authorized the Amazon S3 addon for a project","s3_node_deauthorized":"A user deauthorized the Amazon S3 addon for a project","zotero_folder_selected":"A user linked a Zotero folder to a project","zotero_node_authorized":"A user authorized the Zotero addon for a project","zotero_node_deauthorized":"A user deauthorized the Zotero addon for a project","dataverse_study_released":"A user published a new version of a Dataverse dataset to a project","dataverse_node_deauthorized_no_user":"Dataverse addon for a project deauthorized","affiliated_institution_added":"A user added an affiliation to a project","affiliated_institution_removed":"A user removed an affiliation from a project"}
```
Comment (mfraezz) [inline website/static/js/_anonymousLogTexts.json]: We shouldn't need to commit both this and the `.js` version. `_allLogTexts.json` is .gitignored, this probably should be too.

File: website/static/js/_anonymousLogTexts.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    'made_contributor_visible' : 'A user made contributor(s) visible on a project',
+    'made_contributor_invisible' : 'A user made contributor(s) invisible on a project',
+    'wiki_updated' : 'A user updated a wiki page of a project',
+    'wiki_deleted' : 'A user deleted a wiki page of a project',
+    'wiki_renamed' : 'A user renamed a wiki page of a project',
+    'made_wiki_public' : 'A user made the wiki of a project publicly editable',
+    'made_wiki_private' : 'A user made the wiki of a project privately editable',
+    'addon_added' : 'A user added an addon to a project',
+    'addon_removed' : 'A user removed an addon from a project',
+    'addon_file_moved' : 'A user copied a file',
```
Comment (mfraezz) [inline website/static/js/_anonymousLogTexts.js]: Should read `'A user moved a file'` for consistency with non-anonymized language.

**Comment by mfraezz** [comment]
Would it be possible to implement pagination, rather than arbitrarily extending the log widget?

screen shot 2016-06-10 at 15 49 41

A related bug: when clicking `show more` repeatedly, you end up with the same 'page' of logs repeated. Squint above.

File: website/static/js/_anonymousLogTexts.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    'checked_out':  'A user checked out a file from a project',
+    'checked_in':   'A user checked in a file to a project',
+    'comment_added' : 'A user added a comment to a project',
+    'comment_removed' : 'A user deleted a comment from a project',
+    'comment_updated' : 'A user updated a comment on a project',
+    'embargo_initiated' : 'A user initiated an embargoed registration of a project',
+    'embargo_approved' : 'A user approved an embargoed registration of a project',
+    'embargo_cancelled' : 'A user cancelled an embargoed registration of a project',
+    'embargo_completed' : 'A user completed an embargo of a project',
+    'retraction_initiated' : 'A user A Retraction of a Registration is proposed',
```
Comment (mfraezz) [inline website/static/js/_anonymousLogTexts.js]: Consider instead `A user initiated withdrawal of a registration of a project`

Comment (mfraezz) [inline website/static/js/_anonymousLogTexts.js]: There should also be a `retraction_initiated_no_user` log, reading something like:
>`A withdrawal of a registration of a project was proposed`

Comment (mfraezz) [inline website/static/js/_anonymousLogTexts.js]: There should actually be all of these `_no_user` logs, judging from `logActionList.js`:
```
project_registered_no_user
made_public_no_user
embargo_approved_no_user
embargo_completed_no_user
retraction_initiated_no_user
registration_approved_no_user
```

File: website/static/js/_anonymousLogTexts.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    'retraction_approved' : 'A user approved retraction of registration of a project',
+    'retraction_cancelled' : 'A user cancelled retraction of registration of a project',
```
Comment (mfraezz) [inline website/static/js/_anonymousLogTexts.js]: This and other `retraction_action` logs should have the language updated to replace `retraction` with `withdrawal`.

**Comment by mfraezz** [comment]
Pass complete. Back to you, @caseyrollins 🐊

File: .gitignore [hunk 6]
```diff
@@ -109,6 +109,7 @@ scripts/guid_blacklist.txt
 
 # Generated json
 website/static/js/_allLogTexts.json
+website/static/js/_anonymousLogTexts.json
```
Comment (brianjgeiger) [inline .gitignore]: There's a /website/static/built/ folder that is always hidden; might make sense to use in this case.

**Comment by brianjgeiger** [comment]
Could you update 'after' screenshots with the new functionality?

File: website/static/js/_anonymousLogTexts.js [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    'embargo_approved' : 'A user approved an embargoed registration of a project',
+    'embargo_approved_no_user' : 'Embargo of registration of a project approved',
+    'embargo_cancelled' : 'A user cancelled an embargoed registration of a project',
+    'embargo_completed' : 'A user completed an embargo of a project',
+    'embargo_completed_no_user' : 'Embargo for a project completed',
+    'retraction_initiated' : 'A user initiated a withdrawal of a registration of a project',
+    'retraction_initiated_no_user' : 'A withdrawal of registration of a project is proposed',
+    'retraction_approved' : 'A user approved a withdrawal of registration of a project',
+    'retraction_cancelled' : 'A user cancelled a withdrawal of registration of a project',
+    'registration_initiated' : 'A user initiated a withdrawal of registration of a project',
```
Comment (mfraezz) [inline website/static/js/_anonymousLogTexts.js]: This log message is incorrect.

**Comment by caseyrollins** [comment]
@brianjgeiger, is [this](https://github.com/CenterForOpenScience/osf.io/pull/5825/files#diff-9935e4b6316b40e4e97d56393d0d7f57L145) the Raven message you wanted to be improved? If so, do you have a language suggestion?

**Comment by brianjgeiger** [comment]
Looks like it. See https://openscience.atlassian.net/browse/OSF-6454 (and feel free to close 6454 when you fix it here)

File: website/static/js/_anonymousLogTexts.js [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    'registration_cancelled' : 'A user cancelled registration of a project',
+    'node_created' : 'A user created a project',
+    'node_forked' : 'A user created fork from a project',
+    'node_removed' : 'A user removed a project',
+    'license_changed' : 'A user updated the license of a project',
+    'osf_storage_file_added' : 'A user added a file to OSF Storage in a project',
+    'osf_storage_folder_created' : 'A user created a folder in OSF Storage in a project',
+    'osf_storage_file_updated' : 'A user updated a file in OSF Storage in a project',
+    'osf_storage_file_removed' : 'A user removed a file from OSF Storage in a project',
+    'box_file_added' : 'A user added a file to Box in a project',
```
Comment (mfraezz) [inline website/static/js/_anonymousLogTexts.js]: Please create a follow-up ticket for this if not implemented

File: website/static/js/_anonymousLogTexts.js [hunk 9]
```diff
@@ -0,0 +1,132 @@
+var anonymousLogActions = {
```
Comment (mfraezz) [inline website/static/js/_anonymousLogTexts.js]: Minor: it would be more consistent to name this file `anonymousLogActionList.js`, but this doesn't matter much.

Comment (mfraezz) [inline website/static/js/_anonymousLogTexts.js]: It would also be more consistent for this to be a `.json` file. `logActionsList.js` appears to be unused code left over from the refactor; it's actually `logActionsList.json` that's being used.

This would also obviate writing this to a built `.json` file.

File: webpack.common.config.js [hunk 10]
```diff
@@ -91,6 +92,7 @@ var entry = {
 
 // Collect log text from addons
 var mainLogs = require(staticPath('js/logActionsList.json'));
+var anonymousLogs = require(staticPath('js/_anonymousLogTexts.js'));
 var addonLog;
 
 // Collect addons endpoints. If an addon's static folder has
```
Comment (mfraezz) [inline webpack.common.config.js]: Note: regarding [this comment](https://github.com/CenterForOpenScience/osf.io/pull/5825/files#r66667783)

>Consider refactoring to have the logs for each addon [in addon folders]

Implementing this should be a simple modification to the nested `forEach` loops below here (plus putting the logs in the right places).

File: website/static/js/_anonymousLogTexts.js [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    'osf_storage_folder_created' : 'A user created a folder in OSF Storage in a project',
+    'osf_storage_file_updated' : 'A user updated a file in OSF Storage in a project',
+    'osf_storage_file_removed' : 'A user removed a file from OSF Storage in a project',
+    'box_file_added' : 'A user added a file to Box in a project',
+    'box_file_removed' : 'A user removed a file from Box in a project',
+    'box_file_updated' : 'A user updated a file in Box in a project',
+    'box_folder_created' : 'A user created a folder in Box in a project',
+    'box_folder_selected' : 'A user linked a Box folder to a project',
+    'box_node_authorized' : 'A user authorized the Box addon for a project',
+    'box_node_deauthorized' : 'A user deauthorized the Box addon for a project',
```
Comment (mfraezz) [inline website/static/js/_anonymousLogTexts.js]: This and other `<addon>_node_deauthorized` need a `_no_user` variant:

```
/Users/matt/Devel/osf.io/website/static/js/_allLogTexts.json:
    3: "project_registered_no_user":"${node} registered",
   11: "made_public_no_user":"${node} made public",
   51: "embargo_approved_no_user":"Embargo of registration of ${node} approved",
   54: "embargo_completed_no_user":"Embargo for ${node} completed",
   57: "retraction_initiated_no_user":"A withdrawal of registration of ${node} is proposed",
   62: "registration_approved_no_user":"Registration of ${node} approved",
*  83: "box_node_deauthorized_no_user":"Box addon for ${node} deauthorized",
   92: "dataverse_node_deauthorized_no_user":"Dataverse addon for ${node} deauthorized",
* 100: "dropbox_node_deauthorized_no_user":"Dropbox addon for ${node} deauthorized",
* 107: "figshare_node_deauthorized_no_user":"figshare addon for ${node} deauthorized",
* 115: "github_node_deauthorized_no_user":"GitHub addon for ${node} deauthorized",
* 131: "s3_node_deauthorized_no_user":"Amazon S3 addon for ${node} deauthorized",
* 139: "googledrive_node_deauthorized_no_user":"Google Drive addon for ${node} deauthorized"}

/Users/matt/Devel/osf.io/website/static/js/_anonymousLogTexts.json:
    3: "project_registered_no_user":"Project registered",
   10: "made_public_no_user":"Project made public",
   46: "embargo_approved_no_user":"Embargo of registration of a project approved",
   49: "embargo_completed_no_user":"Embargo for a project completed",
   51: "retraction_initiated_no_user":"A withdrawal of registration of a project is proposed",
   56: "registration_approved_no_user":"Registration of a project approved",
  127: "dataverse_node_deauthorized_no_user":"Dataverse addon for a project deauthorized",
```

File: website/static/js/logTextParser.js [hunk 12]
```diff
@@ -142,7 +147,9 @@ var LogText = {
                 })
             ]);
         } else {
-            message = 'There is no text entry in dictionary for the action :' + logObject.attributes.action;
+            message = 'The log viewer has encountered an unexpected log action: ' + logObject.attributes.action +
+                '. Please add a new log entry for this action to logActionsList.js, ' +
```
Comment (mfraezz) [inline website/static/js/logTextParser.js]: Should be `logActionsList.json`. Not the first time I've seen that cause confusion; it may be best to delete the unused `.js` file.

File: website/static/js/logFeed.js (hunk 1 above)
Comment (mfraezz) [inline website/static/js/logFeed.js]: The sheer number of nested ternaries/conditionals in this would usually be a red flag, but given the complexity of what this is doing, it may be acceptable. I'll leave the final word on that up to @sloria

File: .gitignore (hunk 6 above)
Comment (mfraezz) [inline .gitignore]: I agree, but wouldn't consider this a blocker. It would be a good thing to do in the clean-up PR to remove all the code from the old logging system.

**Comment by mfraezz** [comment]
A couple of non-blocking concerns, but otherwise LGTM 🔥

Passing this along

**Comment by sloria** [comment]
Discussed on Flowdock: there are unwanted grey borders on the project overview pages.

osf___waka

File: website/static/js/logFeed.js [hunk 13]
```diff
@@ -1,238 +1,245 @@
...
+                });
+
+                self.activityLogs(result.data);  // Set activity log data
+                self.nextPage = result.links.next;
+                self.prevPage = result.links.prev;
+                self.firstPage = result.links.first;
+                self.lastPage = result.links.last;
+
+                var params = $osf.urlParams(url);
+                var page = params.page ? params.page : 1;
```
Comment (sloria) [inline website/static/js/logFeed.js]: Very minor: this can be simplified to 

```js
var page = params.page || 1;
```

File: website/static/js/logFeed.js [hunk 14]
```diff
@@ -1,238 +1,245 @@
...
-////////////////
-// Public API //
-////////////////
+        return m('.db-activity-list.m-t-md', [
 
-var defaults = {
-    /** Selector for the progress bar. */
-    progBar: '#logProgressBar'
-};
+            ctrl.failed ? m('p', 'Unable to retrieve logs at this time.') :
```
Comment (sloria) [inline website/static/js/logFeed.js]: For consistency with other error messages on the site, it would be good to include a link to mail OSF support.

File: website/static/js/logFeed.js [hunk 15]
```diff
@@ -1,238 +1,245 @@
...
+                m('.logo-spin.logo-lg'),
+                m('p.m-t-sm.fg-load-message', 'Loading logs...')
+            ]) :
 
-var initViewModel = function(self, logs, url){
-    self.logs = createLogs(logs);
-    self.viewModel = new LogsViewModel(self.logs, url);
-    if(url) {
-        self.viewModel.fetchResults();
+            [ctrl.activityLogs() ? ctrl.activityLogs().map(function(item) {
```
Comment (sloria) [inline website/static/js/logFeed.js]: Minor: No need to nest an array within an array here:

```js
ctrl.activityLogs() ? ctrl.activityLogs().map(...)
```

Comment (sloria) [inline website/static/js/logFeed.js]: Erm, scratch that; I see now there are multiple children nested under `.db-activity-list`, so the outer array may be necessary.

File: website/static/js/logFeed.js [hunk 16]
```diff
@@ -1,238 +1,245 @@
...
+                var image = m('i.fa.fa-desktop');
+                if (ctrl.node.anonymous) { item.anonymous = true; }
+                if (!item.anonymous && item.embeds.user && item.embeds.user.data) {
+                    image = m('img', { src : item.embeds.user.data.links.profile_image});
```
Comment (sloria) [inline website/static/js/logFeed.js]: The image is being requested at the default size (40x40px, I believe), which is much larger than necessary. Better to request a smaller image (16x16) from Gravatar. Check out the Gravatar docs.

I realize this is an issue introduced by the myProjects, so this isn't a blocking change. But I'd suggest making it, since it should be a very small change.

**Comment by sloria** [comment]
Now that `LogFeed` is a proper mithril component, it should be moved to `static/js/components/logFeed.js`.

File: website/static/js/logFeed.js [hunk 17]
```diff
@@ -1,238 +1,245 @@
...
+                    var message = 'Error retrieving logs for ' + self.node.id;
+                    Raven.captureMessage(message, {extra: {url: url, textStatus: textStatus, error: error}});
+                }
+            );
+        };
+
+        self.getCurrentLogs = function _getCurrentLogs (node, page){
+            if(!self.logRequestPending()) {
+                var url = _buildLogUrl(node, page, self.limitLogs);
+                return self.getLogs(url);
```
Comment (sloria) [inline website/static/js/logFeed.js]: Is it possible for `url` to be `undefined` here? It looks like `_buildLogUrl` will return `undefined` for nodes that aren't retracted/withdrawn.

File: website/static/js/logFeed.js [hunk 18]
```diff
@@ -1,238 +1,245 @@
...
+    xhr.withCredentials = window.contextVars.isOnRootDomain;
+    xhr.setRequestHeader('Content-Type', 'application/vnd.api+json;');
+    xhr.setRequestHeader('Accept', 'application/vnd.api+json; ext=bulk');
+};
+
+var _buildLogUrl = function(node, page, limitLogs) {
+    if (!node.isRetracted || !node.is_retracted) {
+        var urlPrefix = (node.isRegistration || node.is_registration) ? 'registrations' : 'nodes';
+        var size = limitLogs ? LOG_PAGE_SIZE_LIMITED : LOG_PAGE_SIZE;
+        var query = { 'page[size]': size, 'page': page, 'embed': ['original_node', 'user', 'linked_node', 'template_node']};
```
Comment (sloria) [inline website/static/js/logFeed.js]: Does our API gracefully handle `page=undefined`? It looks like undefined will be passed [here](https://github.com/CenterForOpenScience/osf.io/blob/938ecbdff7bc53c19bfa87c1cce8b5412e830b1c/website/static/js/logFeed.js#L93). 

Even if it is handled by the API, it would be worth passing `page=1` if page isn't passed, for readability.

```js
page = page || 1;
```

File: website/static/js/logFeed.js [hunk 19]
```diff
@@ -1,238 +1,245 @@
...
-            registrationCancelled: item.node.is_registration && item.node.registered_from_id == null
-        });
-    });
-    return mappedLogs;
-};
+    view : function (ctrl) {
+
+        var i;
+        ctrl.paginators([]);
+        if (ctrl.totalPages > 1 && !ctrl.limitLogs) {
```
Comment (sloria) [inline website/static/js/logFeed.js]: `totalPages` is an m.prop, so you must invoke it to access it.

File: website/static/js/logFeed.js [hunk 20]
```diff
@@ -1,238 +1,245 @@
...
+            // previous page
+            ctrl.paginators().push({
+                url: function() { return ctrl.prevPage; },
```
Comment (sloria) [inline website/static/js/logFeed.js]: Same here; `prevPage` is an m.prop, so you need to invoke it.

```js
url: function() { return ctrl.prevPage(); }
// OR
url: ctrl.prevPage
```

File: website/static/js/logFeed.js [hunk 21]
```diff
@@ -1,238 +1,245 @@
...
+                self.activityLogs(result.data);  // Set activity log data
+                self.nextPage = result.links.next;
+                self.prevPage = result.links.prev;
+                self.firstPage = result.links.first;
+                self.lastPage = result.links.last;
+
+                var params = $osf.urlParams(url);
+                var page = params.page ? params.page : 1;
+                self.currentPage(parseInt(page));
+                self.totalPages = Math.ceil(result.links.meta.total / result.links.meta.per_page);
```
Comment (caseyrollins) [inline website/static/js/logFeed.js]: Ah, actually ```totalPages``` doesn't need to be an m.prop, so I'll change it to have some default value (in which case the way it's assigned and invoked in the rest of the file will be correct).

Comment (caseyrollins) [inline website/static/js/logFeed.js]: Same goes for ```nextPage```, ```prevPage```, ```firstPage```. and ```lastPage``` too.

File: website/static/js/logFeed.js (hunk 1 above)
Comment (sloria) [inline website/static/js/logFeed.js]: Lots of nested ternaries are somewhat expected when writing components, so I think this is ok. Docs certainly don't hurt, though.

File: website/templates/util/render_node.mako [hunk 22]
```diff
@@ -157,3 +137,14 @@
 
 % endif
 </div>
+<link rel="stylesheet" href="/static/css/my-projects.css">
```
Comment (sloria) [inline website/templates/util/render_node.mako]: See previous comment re: my-projects.css

File: website/templates/util/render_node.mako [hunk 23]
```diff
@@ -157,3 +137,14 @@
...
+<script type="text/javascript">
+            window.contextVars = $.extend(true, {}, window.contextVars, {
+            id: ${summary['primary_id'] if not summary['primary'] and summary['can_view'] else summary['id'] | sjson, n},
+            node: ${summary | sjson, n},
+            user: {
+            canView: ${summary['can_view'] | sjson, n}
+    }
+    });
+</script>
+<script src=${"/static/public/js/render-node.js" | webpack_asset}></script>
```
Comment (sloria) [inline website/templates/util/render_node.mako]: Won't this make this script load for every node in the list?

**Comment by sloria** [comment]
Pass finished.

File: website/static/js/mithrilHelpers.js [hunk 24]
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+var $ = require('jquery');  // jQuery
+
+/* Send with ajax calls to work with api2 */
+var apiV2Config = function (xhr) {
+    xhr.withCredentials = window.contextVars.isOnRootDomain;
```
Comment (sloria) [inline website/static/js/mithrilHelpers.js]: We want to avoid using `window` in reuable modules like this. 

See my suggested API for this function here: https://github.com/CenterForOpenScience/osf.io/pull/5825#discussion_r67696208 .

**Comment by sloria** [comment]
I'm getting an error when I build assets 

```

ERROR in Entry module not found: Error: Cannot resolve 'file' or 'directory' /Users/sloria/projects/cos-projects/osf/website/static/js/pages/render-node.js in /Users/sloria/projects/cos-projects/osf
```

I think you have to remove `render-node.js` from the webpack config.

**Comment by caseyrollins** [comment]
Whoops. Removing that now.

**Comment by sloria** [comment]
The log text and date are misaligned

osf___waka

**Comment by caseyrollins** [comment]
@sloria, Nope, the log date is supposed to be below the log text. See the logs on the My Projects page for reference.

**Comment by sloria** [comment]
Oh ok. Thanks for the clarification.

File: website/static/js/logTextParser.js [hunk 25]
```diff
@@ -142,7 +147,10 @@ var LogText = {
                 })
             ]);
         } else {
-            message = 'There is no text entry in dictionary for the action :' + logObject.attributes.action;
+            message = 'The log viewer has encountered an unexpected log action: ' + logObject.attributes.action +
+                '. Please add a new log entry for this action to logActionsList.json' +
```
Comment (sloria) [inline website/static/js/logTextParser.js]: 👍 👍

**Comment by sloria** [comment]
Pass finished.

File: website/static/js/anonymousLogActionsList.json [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+    "made_wiki_private" : "A user made the wiki of a project privately editable",
+    "addon_added" : "A user added an addon to a project",
+    "addon_removed" : "A user removed an addon from a project",
+    "addon_file_moved" : "A user moved a file",
+    "addon_file_copied" : "A user copied a file",
+    "addon_file_renamed" : "A user renamed a file",
+    "folder_created" : "A user created a folder in a project",
+    "file_added" : "A user added a file to a project",
+    "file_updated" : "A user updated a file in a project",
+    "file_removed" : "A user removed a file from a project",
```
Comment (sloria) [inline website/static/js/anonymousLogActionsList.json]: Has the text changed for these anonymized logs? @saradbowman, are we ok with displaying "a project" even for components? If so, please document this in the ticket so QA doesn't report it as a bug.

File: website/static/js/components/logFeed.js [hunk 27]
```diff
@@ -0,0 +1,243 @@
...
+            self.activityLogs([]); // Empty logs from other projects while load is happening;
+            function _processResults (result){
+                result.data.map(function(log){
+                    log.attributes.formattableDate = new $osf.FormattableDate(log.attributes.date);
+                });
+
+                self.activityLogs(result.data);  // Set activity log data
+                self.nextPage(result.links.next);
+                self.prevPage(result.links.prev);
+                self.firstPage = result.links.first;
```
Comment (sloria) [inline website/static/js/components/logFeed.js]: Sorry to contradict what we had talked about in person, but `firstPage` and `lastPage` should actually be m.props because they do change over time. They are initialized in the controller, then they change after the first log request is made.

Making them props ensures that the view will update when they are updated. This is working now  only because `nextPage` and `prevPage` are also updated in the same callback.

File: website/static/js/components/logFeed.js [hunk 28]
```diff
@@ -0,0 +1,243 @@
...
+        var self = this;
+        self.node = options.node;
+        self.activityLogs = m.prop();
+        self.logRequestPending = m.prop();
+        self.limitLogs = options.limitLogs;
+        self.failed = false;
+        self.paginators = m.prop([]);
+        self.nextPage = m.prop();
+        self.prevPage = m.prop();
+        self.firstPage = 0;
```
Comment (sloria) [inline website/static/js/components/logFeed.js]: It is misleading to initialize these to an integer because they are eventually assigned as strings (links). As mentioned in my previous comments, these should be m.props, initialized as a string (`m.prop('')`).

File: website/static/js/components/logFeed.js (hunk 27 above)
Comment (sloria) [inline website/static/js/components/logFeed.js]: Same with `totalPages`.

File: website/static/js/components/logFeed.js [hunk 29]
```diff
@@ -0,0 +1,243 @@
...
+            }
+            self.logRequestPending(true);
+            var promise = m.request({method : 'GET', url : url, config: mHelpers.apiV2Config({withCredentials: window.contextVars.isOnRootDomain})});
+            promise.then(
+                function(result) {
+                    _processResults(result);
+                    self.logRequestPending(false);
+                    return promise;
+                }, function(xhr, textStatus, error) {
+                    self.failed = true;
```
Comment (sloria) [inline website/static/js/components/logFeed.js]: Shouldn't `logRequestPending` be set to `true` here?

File: website/static/js/anonymousLogActionsList.json (hunk 26 above)
Comment (caseyrollins) [inline website/static/js/anonymousLogActionsList.json]: The anonymized logs were demoed to @saradbowman in this way. Since the API doesn't return any information about the node type when you request with a VOL, we can either display ```a project``` for all logs, or nothing (i.e. "A user added a file" instead of "A user added a file to a project"). @saradbowman, thoughts?

Comment (brianjgeiger) [inline website/static/js/anonymousLogActionsList.json]: You could take off "from a project" entirely. Or we could leave it and file a future improvement ticket.

File: website/static/js/components/logFeed.js [hunk 30]
```diff
@@ -0,0 +1,243 @@
...
+            ]) :
+            // Show OSF spinner while there is a pending log request
+            ctrl.logRequestPending() ?  m('.spinner-loading-wrapper', [
+                m('.logo-spin.logo-lg'),
+                m('p.m-t-sm.fg-load-message', 'Loading logs...')
+            ]) :
+            // Display each log item (text and user image)
+            [ctrl.activityLogs() ? ctrl.activityLogs().map(function(item) {
+                var image = m('i.fa.fa-desktop');
+                if (ctrl.node.anonymous) { item.anonymous = true; }
```
Comment (sloria) [inline website/static/js/components/logFeed.js]: Why is this necessary? Can you just check `ctrl.node.anonymous` in the line below?

File: website/static/js/components/logFeed.js (hunk 27 above)
Comment (caseyrollins) [inline website/static/js/components/logFeed.js]: Just realized I'm not using ```firstPage``` and ```lastPage``` anywhere so they can actually be removed.

**Comment by sloria** [comment]
Pass finished. Just tiny changes left. 👪

File: website/static/js/components/logFeed.js (hunk 30 above)
Comment (caseyrollins) [inline website/static/js/components/logFeed.js]: Good catch. It's completely unnecessary.

**Comment by caseyrollins** [comment]
@sloria :shipit:

**Comment by sloria** [comment]
This is ready to merge. Will merge this after 0.78.0 is released.
